File: app/src/main/java/com/pr0gramm/app/io/Cache.java
package com.pr0gramm.app.io;

import android.content.Context;
import android.net.Uri;

import com.google.common.collect.Iterables;
import com.google.common.collect.Ordering;
import com.pr0gramm.app.BuildConfig;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.Closeable;
import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.TimeUnit;

import javax.inject.Inject;
import javax.inject.Singleton;

import okhttp3.OkHttpClient;
import rx.Observable;
import rx.schedulers.Schedulers;

import static com.pr0gramm.app.util.AndroidUtility.toFile;

/**
 * A cache we can use for linear caching of http requests.
 */
@Singleton
public class Cache {
    private final long MEGA = 1024 * 1024;

    private final Logger logger = LoggerFactory.getLogger("Cache");

    private final Object lock = new Object();
    private final Map<String, Entry> cache = new HashMap<>();

    private final long maxCacheSize;
    private final File root;
    private final OkHttpClient httpClient;

    @Inject
    public Cache(Context context, OkHttpClient httpClient) {
        this.root = new File(context.getCacheDir(), "mediacache");
        this.httpClient = httpClient;

        // schedule periodic cache clean up.
        Observable
                .interval(10, 60, TimeUnit.SECONDS, Schedulers.io())
                .doOnNext(v -> cleanupCache())
                .onErrorResumeNext(Observable.empty())
                .subscribe();

        // print cache every few seconds for debugging
        if (BuildConfig.DEBUG) {
            Observable
                    .interval(5, TimeUnit.SECONDS, Schedulers.io())
                    .subscribe(v -> printCache());
        }

        this.maxCacheSize = (root.getFreeSpace() > 1024 * MEGA ? (256 * MEGA) : (128 * MEGA));
        logger.debug("Initialized cache with {}mb of space", maxCacheSize / (double) MEGA);
    }

    private void printCache() {
        synchronized (cache) {
            logger.debug("Cache:");
            for (Entry entry : cache.values()) {
                logger.debug("  * {}", entry);
            }
        }
    }

    /**
     * Returns a cached or caching entry. You need to close your reference
     * once you are finish with it.
     */
    public Cache.Entry get(Uri uri) {
        String key = uri.toString();

        synchronized (lock) {
            Entry entry = cache.get(key);
            if (entry == null) {
                entry = createEntry(uri);
                cache.put(key, entry);
            }

            return refCountIfNeeded(entry);
        }
    }

    private Entry refCountIfNeeded(Entry entry) {
        if (entry instanceof CacheEntry) {
            ((CacheEntry) entry).incrementRefCount();
        }

        return entry;
    }

    private Entry createEntry(Uri uri) {
        logger.debug("Creating a new cache entry for uri {}", uri);

        // just open the file directly if it is local.
        if ("file".equals(uri.getScheme())) {
            return new FileEntry(toFile(uri));
        }

        File cacheFile = cacheFileFor(uri);
        return new CacheEntry(httpClient, cacheFile, uri);
    }

    /**
     * Retuns the cache file for the given uri.
     */
    private File cacheFileFor(Uri uri) {
        String filename = uri.toString().replaceFirst("https?://", "").replaceAll("[^a-zA-Z0-9.]+", "_");
        return new File(root, filename);
    }

    /**
     * Checks for old files in the cache directory and removes the oldest files
     * if they exceed the maximum cache size.
     */
    private void cleanupCache() {
        if (!root.exists()) {
            return;
        }

        List<File> files = Ordering.natural()
                .onResultOf(File::lastModified)
                .reverse()
                .sortedCopy(Arrays.asList(root.listFiles()));

        logger.debug("Doing cache cleanup, found {} files", files.size());

        long totalSize = 0;
        for (File file : files) {
            totalSize += file.length();

            if (totalSize > maxCacheSize) {
                forgetEntryForFile(file);
            }
        }

        logger.debug("Cache took {}mb", totalSize / (1024.f * 1024.f));
    }

    /**
     * Removes the cached entyr with the given filename.
     */
    private void forgetEntryForFile(File file) {
        logger.debug("Remove old cache file {}", file);
        if (!file.delete()) {
            logger.warn("Could not delete cached file {}", file);
        }

        synchronized (lock) {
            for (Entry entry : cache.values()) {
                if (entry instanceof CacheEntry) {
                    if (((CacheEntry) entry).deleteIfClosed()) {
                        // remove the entry from our cache
                        Iterables.removeIf(cache.values(), entry::equals);
                        break;
                    }
                }
            }
        }
    }

    public interface Entry extends Closeable {
        int totalSize() throws IOException;

        InputStream inputStreamAt(int offset) throws IOException;

        /**
         * Returns a value between 0 and 1 that specifies how much of this
         * entry is actually cached. Returns -1 if no estimate is currently available.
         */
        float fractionCached();

        @Override
        void close();
    }
}

File: app/src/main/java/com/pr0gramm/app/io/Cache.kt
package com.pr0gramm.app.io

import android.app.Application
import android.net.Uri
import androidx.core.net.toFile
import com.pr0gramm.app.Duration.Companion.seconds
import com.pr0gramm.app.Logger
import com.pr0gramm.app.Stats
import com.pr0gramm.app.util.doInBackground
import com.pr0gramm.app.util.isLocalFile
import com.pr0gramm.app.util.runEvery
import okhttp3.MediaType
import okhttp3.OkHttpClient
import okhttp3.Protocol
import okhttp3.Request
import okhttp3.Response
import okhttp3.ResponseBody.Companion.asResponseBody
import okio.buffer
import okio.source
import java.io.Closeable
import java.io.File
import java.io.InputStream


/**
 * A cache we can use for linear caching of http requests.
 */
class Cache(private val context: Application, private val httpClient: OkHttpClient) {
    private val Long.mega: Long get() = this * 1024L * 1024L

    private val logger = Logger("Cache")
    private val root: File = File(context.cacheDir, "mediacache")

    private val lock = Any()
    private val cache = HashMap<String, Entry>()

    init {
        doInBackground {
            runEvery(seconds(60), initial = seconds(10)) {
                cleanupCache()
            }
        }
    }

    /**
     * Returns a cached or caching entry. You need to close your reference
     * once you are finish with it.
     */
    fun get(uri: Uri): Entry {
        val key = cacheKeyOf(uri)

        synchronized(lock) {
            logger.debug { "Looking up cache entry for $key" }

            var entry: Entry? = cache[key]
            if (entry?.valid != true) {
                entry = createEntry(uri)
                cache[key] = entry
            }

            return refCountIfNeeded(entry)
        }
    }

    private fun refCountIfNeeded(entry: Entry): Entry {
        if (entry is CacheEntry) {
            entry.incrementRefCount()
        }

        return entry
    }

    private fun createEntry(uri: Uri): Entry {
        logger.debug { "Creating a new cache entry for uri $uri" }

        // just open the file directly if it is already local.
        if (uri.isLocalFile) {
            logger.debug { "Uri is local, returning FileEntry(${uri.toFile()})" }
            return FileEntry(uri.toFile())
        }

        // derive the files basename from the url
        val basename = uri.toString()
                .replaceFirst("https?://".toRegex(), "")
                .replace("[^a-zA-Z0-9.]+".toRegex(), "_")

        // check if we have a fully cached file
        val fullyCached = File(root, "$basename.ok")
        if (fullyCached.exists() && fullyCached.length() > 0) {
            logger.debug { "Fully cached file exists, returning FileEntry(${fullyCached}" }
            return FileEntry(fullyCached)
        }

        logger.debug { "No cached file found for $uri, creating new CacheEntry." }

        // resume a previously cached file
        val partialCached = File(root, "$basename.part")
        return CacheEntry(httpClient, partialCached, fullyCached, uri)
    }

    /**
     * Checks for old files in the cache directory and removes the oldest files
     * if they exceed the maximum cache size.
     */
    private fun cleanupCache() {
        if (!root.exists()) {
            return
        }

        val filesInRoot = root.listFiles() ?: return

        // cache times to sort without someone modifying timestamps during sorting between
        val modificationTimes = filesInRoot.associate { Pair(it, it.lastModified()) }
        val files = modificationTimes.keys.sortedByDescending { modificationTimes[it] }

        // The space already in use by the cache
        val usedCacheSpace = files.fold(0L) { acc, file -> acc + file.length() }

        // The amount that can be used by apps
        val usableSpace = root.usableSpace - 1024L.mega

        // Ignoring what the cache already uses, this is the amount of space that
        // is available to us.
        val availableSpace = usedCacheSpace + usableSpace

        // Now put it in sane limits.
        val maxCacheSize = availableSpace.coerceIn(256L.mega, 512L.mega)

        logger.debug { "Doing cache cleanup, maxCacheSize=${formatSpace(maxCacheSize)}, found ${files.size} files using ${formatSpace(usedCacheSpace)}" }

        var usedSpace: Long = 0
        for (file in files) {
            usedSpace += file.length()

            if (usedSpace > maxCacheSize) {
                forgetEntryForFile(file)
            }
        }

        // do some tracking of cache sizes
        Stats().histogram("cache.maxSize", maxCacheSize)
        Stats().histogram("cache.usedSize", usedSpace)
        Stats().histogram("cache.usage", usedSpace.toDouble() / maxCacheSize)
    }

    /**
     * Removes the cached entry with the given filename.
     */
    private fun forgetEntryForFile(file: File) {
        logger.debug { "Remove old cache file $file" }
        if (!file.delete()) {
            logger.warn { "Could not delete cached file $file" }
        }
    }

    private fun formatSpace(space: Long): String {
        return "%1.2fmb".format(space.toFloat() / (1024f * 1024f))
    }

    private fun cacheKeyOf(uri: Uri): String {
        return uri.toString().replaceFirst("http://", "https://")
    }

    interface Entry : Closeable {
        /**
         * Total size of this cache entry.
         */
        val totalSize: Int

        /**
         * Returns a value between 0 and 1 that specifies how much of this
         * entry is actually cached. Returns -1 if no estimate is currently available.
         */
        val fractionCached: Float

        /**
         * Not null if this cache entry is backed by a file.
         */
        val file: File?

        /**
         * Check if the entry is still valid
         */
        val valid: Boolean get() = file?.exists() == true

        /**
         * An input stream that starts at the given place in the file.
         */
        fun inputStreamAt(offset: Int): InputStream

        /**
         * Closes the reference to the entry. Must be called after getting
         * a Entry instance using Cache#get.
         */
        override fun close()

        fun toResponse(request: Request, mediaType: MediaType? = null): Response {
            val body = inputStreamAt(0).source().buffer().asResponseBody(mediaType, totalSize.toLong())

            return Response.Builder()
                    .request(request)
                    .protocol(Protocol.HTTP_1_0)
                    .code(200)
                    .message("OK")
                    .body(body)
                    .build()
        }
    }
}
